rfcs/20230503-profiling-api/README.md

```diff
@@ -0,0 +1,196 @@
+cases to reliably define the semantics. The API can be enabled with a build
+time option `ONEDNN_EXPERIMENTAL_PROFILING`.
+
+The profiling API will have the following requirements:
+* The profiling mechanism will have to be enabled and disabled explicitly with
+the corresponding API
+* When the mechanism is enabled the oneDNN stream is expected to contain a queue
+that was created with enabled profiling capabilities
+* Explicit synchronization after submitting a primitive to the queue is required
+to query the profiling data for the primitive
+
+In order to support profiling for multiple streams each stream will have its own
+profiler, this is why the profiling API will have a stream as a parameter.
+
+A typical workflow looks as follows:
+* Enable profiling
+* Create a oneDNN stream with a queue that was created with enabled profiling
+capabilities
+* Reset the profiler's state before executing the primitive
+* Execute the primitive and explicitly wait for it to complete
+* Use the corresponding API to query profiling data
+* Reset the profiler's state
+* Disable profiling
+
+### API
+
+The API for enabling and disabling profiling:
+
+#### C
+```c
+/// Enable profiling capabilities.
+///
+/// @param stream Stream to be used for profiling.
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t dnnl_enable_profiling(dnnl_stream_t stream);
+
+/// Disable profiling capabilities.
+///
+/// @param stream Stream that is used for profiling
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t dnnl_disable_profiling(dnnl_stream_t stream);
+```
+
+#### C++
+```cpp
+/// @copydoc dnnl_enable_profiling()
+status enable_profiling(stream &stream);
+
+/// @copydoc dnnl_disable_profiling()
+status disable_profiling(stream &stream);
+```
+
+The API for resetting the profiler's state:
+
+#### C
+```c
+/// Resets a profiler's state.
+///
+/// @param stream Stream associated with the profiler.
+///
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t dnnl_reset_profiling(dnnl_stream_t stream);
+```
+
+#### C++
+```cpp
+/// @copydoc dnnl_reset_profiling()
+status reset_profiling(stream &stream);
+```
+
+The API to query profiling data.
+
+#### C
+
+```c
+/// Profiling data kind.
+typedef enum {
+    /// Undefined profiling data kind.
+    dnnl_profiling_data_kind_undef = 0,
+    /// Data kind to query an execution time in nanoseconds.
+    dnnl_profiling_data_kind_time,
+} dnnl_profiling_data_kind_t;
+
+/// Profiling mode.
+typedef enum {
+    /// Undefined profiling mode.
+    dnnl_profiling_mode_undef = 0,
+    /// The profiling data for each kernel will be added up. For example,
+    /// if the profiling data kind is time and there are n kernels then the
+    /// result will be timeK0 + timeK1 + timeKn-1.
+    dnnl_profiling_mode_sum,
```

Comment:
In the general case oneDNN may execute an arbitrary number of various kernels under `primitive.execute()`, e.g. a lightweight zero initialization kernel followed by a heavy compute kernel. With this it doesn't look useful to provide anything but the total time of all kernels. Do you anticipate any other modes in the future?

Reply:
There is a concern that in the case when the kernels within the same primitive are executed out-of-order a simple addition may not be accurate. This API allows us to add a mode that would work better for the mentioned cases though I'm not sure what the mode should look like at this point.

Reply:
I see, yeah I forgot about out-of-order. We may think of something like `(max_time_end - min_time_start)` to rely on the device-side wall time across all kernel submissions.

Reply:
Wouldn't `(max_time_end - min_time_start)` include host overhead? This seems to be very similar to the markers approach.

Reply:
That's true but I can't think of something more useful for out-of-order execution. Time of a single kernel is not relevant anymore because of potential overlapping.

Reply:
I agree on the out-of-order and multiple-in-order queue lack of usefulness:
- sum of kernel time do not provide much information as multiple kernel (within a primitive, or across streams) can run sequentially or overlap.
- if kernels execution within a primitive do not overlap, or if kernel execution time is smaller than host overhead, `time_end - time_start` will not strictly be device time.

IMHO, the best way for a user to profile device time is by following the guidelines [here ](https://github.com/intel/pti-gpu), by intercepting events with [cl_tracer](https://github.com/intel/pti-gpu/tree/master/tools/cl_tracer) APIs.

Anyway, providing the current API seems to be usable enough for OpenVINO, and putting it in experimental will allow us to make it evolve as gaps in metrics collection are discovered.

Reply:
Given that the way the kernels are executed is an implementation detail it would be better to avoid exposing it. I'm OK with removing `dnnl_profiling_mode` so that the library can decide what is the right way to collect the data.